runner_manager/backend/aws.py

```diff
@@ -16,66 +18,130 @@ class AWSBackend(BaseBackend):
-            instance_id = instance["Instances"][0]["InstanceId"]
         except ClientError as e:
-            if e.response["Error"]["Code"] == "InvalidInstanceID.NotFound":
+            error = e.response.get("Error", {})
+            if error.get("Code") == "InvalidInstanceID.NotFound":
                 log.error(f"Instance {instance_id} not found.")
             else:
                 raise e
+        try:
+            reservations = instance.get("Reservations", [])
+            if reservations:
+                instances = reservations[0].get("Instances", [])
+                if instances:
+                    instance_id = instances[0]["InstanceId"]
+                else:
+                    raise ValueError(f"No instances found for {instance_id}")
+            else:
+                raise ValueError(f"No reservations found for {instance_id}")
+        except Exception as e:
+            raise e
         return Runner.find(Runner.instance_id == instance_id).first()
 
     def list(self) -> List[Runner]:
         """List runners."""
         runners: List[Runner] = []
         try:
-            instances = self.client.describe_instances()
-            for instance in instances["Reservations"]:
-                for i in instance["Instances"]:
-                    labels = i["Tags"]
-                    if self.manager and "runner-manager" in labels:
-                        runner = Runner(
-                            name=i["InstanceId"],
-                            instance_id=i["InstanceId"],
-                            busy=False,
-                        )
-                        runners.append(runner)
+            reservations = self.client.describe_instances(
+                Filters=[
+                    {
+                        "Name": "tag-key",
+                        "Values": ["runner-manager"],
+                    },
+                    {
+                        "Name": "instance-state-name",
+                        "Values": ["running"],
+                    },
+                ]
+            ).get("Reservations", [])
+        except Exception as e:
+            raise e
+        for reservation in reservations:
+            for instance in reservation.get("Instances", []):
+                try:
+                    runners.append(
+                        Runner.find(
+                            Runner.instance_id == instance.get("InstanceId", "")
+                        ).first()
+                    )
+                except NotFoundError:
+                    runner = Runner(
+                        name=instance.get("InstanceId", ""),
+                        instance_id=instance.get("InstanceId", ""),
+                        busy=False,
+                    )
+                    runners.append(runner)
+        return runners or []
+
+    def update(self, runner: Runner) -> Runner:
+        """Update a runner."""
+        try:
+            if runner.instance_id:
```

Comment:
Don't put a "if" in a try statement in this case. You want the content of the "try" to be as small as possible

Reply:
ok noted